### PR TITLE
DS-3171: Helper functions for Table - Select

### PR DESCRIPTION
**Pull request by mwmclean**

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # Generated by roxygen2: do not edit by hand
 
+export(DropMultipleStatisticsFromTable)
```
Comment (JustinCCYap) [inline NAMESPACE:3]: I don't think this needs to be exported?

File: NAMESPACE [hunk 2]
```diff
@@ -1,6 +1,9 @@
...
 export(First)
+export(FlattenQTableToMatrix)
```
Comment (JustinCCYap) [inline NAMESPACE:5]: I don't think this needs to be exported either

**Comment by JustinCCYap** [comment]
A PR needs to be created for select.R in Q-Wiki-Scripts to use the function(s) in this package

**Comment by JustinCCYap** [comment]
Is `selectFromTableUsingCheckBoxes` not being moved to verbs?

**Comment by mwmclean** [comment]
`DropMultipleStatisticsFromTable` and `FlattenQTableToMatrix` don't need to be exported, but I don't see the harm in exporting them and we might have more use for them later on.

File: R/select.R [hunk 3]
```diff
@@ -1,60 +1,174 @@
-#' @noRd
-#' Internal functions to select from QTables using Combo Box
-#' R Controls, which are reprene
-#' @return the number of tables supplied by the user using Table - Select GUI controls
-nTables <- function()
-    length(ls(pattern = "^formTables[0-9]+$", envir = .GlobalEnv)) - 1
-
-# Subsets a table based on values of combo.boxes
-selectFromTableUsingComboBoxes <- function(table, table.orig, combo.boxes, dim = 1)
```
Comment (JustinCCYap) [inline R/select.R]: This will break existing outputs for users when it is removed?

Comment (mwmclean) [reply R/select.R]: We haven't released a version of Select that uses `verbs` yet, that function only appears in the Standar R code, so there should be no issues.

File: tests/testthat/test-select.R [hunk 4]
```diff
@@ -91,3 +91,202 @@ test_that("5D QTables",
...
+    expect_equal(nrow(out), 1)
+    expect_equivalent(attr(out, "span")$rows, "18 to 24")
+
+    ridx <- c("Pepsi", "Diet Pepsi", "Pepsi Max")
+    cidx <- c("Older", "Open to new experiences")
+    out <- SelectFromTable(qtable.2D.multi.VS,
+                    row.selection.mode = "vector",
+                    row.selections = ridx,
+                    column.selections = cidx)
+    expect_equal(dim(out), c(length(ridx), length(cidx)))
```
Comment (JustinCCYap) [inline tests/testthat/test-select.R:127]: Add `expect_equivalent(out, qtable.2D.multi.VS[ridx, cidx])` and also similar for the test above?

File: tests/testthat/test-select.R [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+    out <- SelectFromTable(x, row.selection.mode = "vector",
+                           row.selections = LETTERS[row.keep],
+                           column.selection.mode = "last date-time periods",
+                           column.selections = 30, unit = "Day", calendar = FALSE)
+    expect_equal(dim(out), c(2, 2))
+    expect_equivalent(out, Last(x, keep = 30, unit = "Day",
+                                calendar = FALSE)[row.keep, ])
+})
+
+test_that("range selection mode",
```
Comment (JustinCCYap) [inline tests/testthat/test-select.R:244]: Need to add tests for bad inputs, e.g. incorrect format, out of range, out of order

File: R/select.R [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+                           unit = unit, calendar = calendar, ...)
+        selections <- seq_len(nrow(table.out))
+    }else if (grepl("^last", selection.mode)){
+        table.out <- Last(table, keep = selections,
+                          unit = unit, calendar = calendar, ...)
+        selections <- (nrow(table) - nrow(table.out) + 1):nrow(table)
+    }else
+    {
+        if (is.character(selections))
+            selections <- checkSuppliedIndices(table, selections, 1)
```
Comment (JustinCCYap) [inline R/select.R:172]: Need to check that selections are within range otherwise throw nice error

Comment (JustinCCYap) [reply R/select.R:172]: Does order of parsed selections matter or do we just sort numerically? How about duplicate selections?

File: tests/testthat/test-select.R (hunk 5 above)
Comment (JustinCCYap) [reply tests/testthat/test-select.R:244]: Also test what happens with duplicate selections

File: tests/testthat/test-select.R (hunk 4 above)
Comment (mwmclean) [reply tests/testthat/test-select.R:127]: Done

File: tests/testthat/test-select.R (hunk 5 above)
Comment (mwmclean) [reply tests/testthat/test-select.R:244]: I've added the tests. I've also added a call to `unique()` so duplicate selections are now ignored.

File: R/select.R (hunk 6 above)
Comment (mwmclean) [reply R/select.R:172]: I've added a check of the numeric indices/ranges in a function `checkNumericSelections()` with an informative error/warning message if all/any selections are invalid and need to be ignored.

The order does matter. I've made this clear in the unit tests. Duplicate selections are ignored.

File: tests/testthat/test-select.R (hunk 5 above)
Comment (JustinCCYap) [reply tests/testthat/test-select.R:244]: Could you commit the changes as I don't see them yet

File: R/select.R [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+checkNumericSelections <- function(table, indices, dim)
+{
+    indices <- unique(as.integer(indices))
+    if (dim == 1)
+        n <- nrow(table)
+    else
+        n <- ncol(table)
+    bad.idx <- which(indices < 1 | indices > n)
```
Comment (JustinCCYap) [inline R/select.R:280]: Should we also exclude non-integer values and warn when duplicates have been excluded?

File: R/select.R [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+
+    bad.idx <- which(!indices %in% names)
+    if (length(bad.idx))
+    {
+        dim.str <- ifelse(dim == 1, "row", "column")
+        if (length(bad.idx) == length(indices))
+            stop("No valid selections were found in the ", dim.str,
+                 " labels. If supplying selections using a page control, check ",
+                 " the item list for the control and edit if necessary.",
+                 call. = FALSE)
```
Comment (JustinCCYap) [inline R/select.R:261]: This doesn't seem to make a difference? I still see the call details when I tested this

**Comment by JustinCCYap** [comment]
Can we increase the test coverage?

File: R/select.R (hunk 7 above)
Comment (mwmclean) [reply R/select.R]: I have added warnings if duplicates get removed.

File: R/select.R (hunk 8 above)
Comment (mwmclean) [reply R/select.R:276]: Hmm, not sure what's happening there. It's been a documented feature of base R for a long time. Did you try outside RStudio or in Displayr?

**Comment by mwmclean** [comment]
I've added several more tests for parts of the new code that weren't covered, including tables with banners in the rows, data.frames, and some more error handling.

File: R/select.R (hunk 8 above)
Comment (JustinCCYap) [reply R/select.R:276]: In Displayr, maybe Displayr ignores the parameter
